### PR TITLE
Do not add arrays when plotting

### PR DESCRIPTION
**Pull request by akaszynski**
Requires #2754.

As noted in [this comment](https://github.com/pyvista/pyvista/pull/1953#discussion_r890343257) we add arrays when plotting scalars in some cases. For example:

```py
Create a cube and clear its point data.

>>> import pyvista as pv
>>> import numpy as np
>>> cube = pv.Cube()
>>> cube.clear_data()
>>> cube.point_data
pyvista DataSetAttributes
Association     : POINT
Active Scalars  : None
Active Vectors  : None
Active Texture  : None
Active Normals  : None
Contains arrays : None

Plot it. Show that by merely plotting we modify the data.

>>> data = np.random.random(cube.n_points)
>>> cube.plot(scalars=data, off_screen=True)
>>> cube.point_data
pyvista DataSetAttributes
Association     : POINT
Active Scalars  : Data-normed
Active Vectors  : None
Active Texture  : None
Active Normals  : None
Contains arrays :
    Data                    float64    (8,)                 SCALARS

Same thing happens with multi-component data.

>>> cube.clear_data()
>>> data = np.random.random((cube.n_points, 3))
>>> cube.plot(scalars=data, off_screen=True)
>>> cube.point_data
pyvista DataSetAttributes
Association     : POINT
Active Scalars  : Data-normed
Active Vectors  : None
Active Texture  : None
Active Normals  : None
Contains arrays :
    Data-normed             float64    (8,)                 SCALARS
```

This is a bad practice, leads to additional memory bload when plotting, and sometimes even changes the active scalars. This PR fixes this issue by removing any added scalars when the plotter is closed (either by manually calling `close()` or when `show()` is called.

### CODE REVIEW

**Comment by akaszynski** [comment]
Hold off review of this until #2754 is merged.

File: pyvista/_vtk.py [hunk 1]
```diff
@@ -111,16 +111,34 @@
...
+        vtkTypeUInt8Array,
+        vtkTypeUInt16Array,
         vtkTypeUInt32Array,
+        vtkTypeUInt64Array,
+        vtkUnicodeStringArray,
         vtkUnsignedCharArray,
+        vtkUnsignedIntArray,
+        vtkUnsignedLongArray,
+        vtkUnsignedLongLongArray,
+        vtkUnsignedShortArray,
```
Comment (adeak) [inline pyvista/_vtk.py:141]: I think these additions were left over from one of the other PRs and not necessary.

File: tests/plotting/test_plotting.py [hunk 2]
```diff
@@ -1651,7 +1651,8 @@ def test_opacity_by_array_uncertainty(uniform):
 
 
 def test_opacity_by_array_user_transform(uniform):
-    uniform['Spatial Point Data'] /= uniform['Spatial Point Data'].max()
+    scaled = uniform['Spatial Point Data'] / uniform['Spatial Point Data'].max()
+    uniform['Spatial Point Data'] = scaled
```
Comment (adeak) [inline tests/plotting/test_plotting.py:1655]: Can this be reverted now that we've fixed return type and reference issues?

File: tests/test_plotter.py [hunk 3]
```diff
@@ -90,3 +90,42 @@ def test_prepare_smooth_shading_not_poly(hexbeam):
     )
 
     assert np.allclose(mesh[scalars_name], expected_mesh[scalars_name])
+
+
+def test_get_datasets(sphere, hexbeam):
+    """Test edge cases for smooth shading"""
```
Comment (adeak) [inline tests/test_plotter.py:96]: There doesn't seem to be any smooth shading in this test. Nor obvious edge caseness.

File: pyvista/plotting/mapper.py [hunk 4]
```diff
@@ -224,7 +224,7 @@ def set_scalars(
                 rgb or _custom_opac,
             )
 
-            return show_scalar_bar, n_colors, clim
+            return show_scalar_bar, n_colors, clim, new_scalars
```
Comment (adeak) [inline pyvista/plotting/mapper.py:227]: Is this addition used anywhere? And should we document the return values now? Or is this an internal helper function?

**Review by adeak — COMMENTED** [review]
I've also changed a few things, see also https://github.com/pyvista/pyvista/pull/2818.

File: pyvista/plotting/mapper.py [hunk 5]
```diff
@@ -298,10 +312,20 @@ def configure_scalars_mode(
...
         ):
-            """Set custom opacity."""
+            """Set custom opacity.
+
+            Returns
+            -------
+            str or None
+                If the scalars do not exist within the dataset, this is the
+                name of the scalars array.
+
```
Comment (adeak) [inline pyvista/plotting/mapper.py:327]: This returns a tuple now, looking at `configure_scalars_mode()`.

File: pyvista/plotting/mapper.py [hunk 6]
```diff
@@ -322,9 +346,10 @@ def set_custom_opacity(
             rgba[:, :-1] = Color(color, default_color=theme.color).int_rgb
             rgba[:, -1] = np.around(opacity * 255)
 
-            self.configure_scalars_mode(
+            new_scalars = self.configure_scalars_mode(
```
Comment (adeak) [inline pyvista/plotting/mapper.py:349]: In light of the tuple, the name `new_scalars` here and in `plotting.py` might be misleading.

File: pyvista/plotting/plotting.py [hunk 7]
```diff
@@ -2414,6 +2416,10 @@ def add_mesh(
         else:
             self.mapper.SetScalarModeToUseFieldData()
 
+        # track if any data arrays have been added
+        if new_scalars:
```
Comment (adeak) [inline pyvista/plotting/plotting.py:2420]: This is a bit fishy to me. When we look up scalars by name that are already present in the dataset, we get e.g. `new_scalars = (None, 'point')`. So we would add them to the `_added_scalars`. Of course there's no risk of these causing a problem, because we don't store the original name so we can't accidentally remove it from the dataset.

But shouldn't we do something like
```suggestion
        # track if any data arrays have been added
        if new_scalars is not None and new_scalars[0] is not None:
```
to filter out these cases?

File: pyvista/plotting/plotting.py [hunk 8]
```diff
@@ -4585,6 +4594,31 @@ def export_obj(self, filename):
...
+                if hasattr(mapper, 'GetInputAsDataSet'):
+                    datasets.append(mapper.GetInputAsDataSet())
+
+        return datasets
+
+    def _remove_added_scalars(self):
+        """Remove any scalars added to the datasets."""
+        for mesh, (name, assoc) in self._added_scalars:
+            if name is None:
+                continue
```
Comment (adeak) [inline pyvista/plotting/plotting.py:4615]: If we remove `None` cases before addition we can remove this branch.

File: pyvista/plotting/plotting.py [hunk 9]
```diff
@@ -4585,6 +4594,31 @@ def export_obj(self, filename):
...
+        datasets = []
+        for renderer in self.renderers:
+            for actor in renderer.actors.values():
+                mapper = actor.GetMapper()
+
+                # ignore any mappers whose inputs are not datasets
+                if hasattr(mapper, 'GetInputAsDataSet'):
+                    datasets.append(mapper.GetInputAsDataSet())
+
+        return datasets
```
Comment (adeak) [inline pyvista/plotting/plotting.py:4609]: This only seems to be used in a test, so I assume this is for some other work :P

File: pyvista/_vtk.py (hunk 1 above)
Comment (akaszynski) [reply pyvista/_vtk.py]: Correct. Removed.

File: tests/test_plotter.py (hunk 3 above)
Comment (akaszynski) [reply tests/test_plotter.py]: Quite right.

File: tests/plotting/test_plotting.py (hunk 2 above)
Comment (akaszynski) [reply tests/plotting/test_plotting.py]: Correct. This has been cleaned up.

File: pyvista/plotting/mapper.py (hunk 5 above)
Comment (akaszynski) [reply pyvista/plotting/mapper.py]: Fixed.

File: pyvista/plotting/plotting.py (hunk 9 above)
Comment (akaszynski) [reply pyvista/plotting/plotting.py]: There have been times in the past where this would have been useful as a convenience method. This is indeed used in a closed source project; maybe others will find it useful.

File: pyvista/plotting/plotting.py (hunk 8 above)
Comment (akaszynski) [reply pyvista/plotting/plotting.py]: Resolved in 32fcae6f21b8213e6c988a2d78a960e9a0300d09.

File: pyvista/plotting/plotting.py (hunk 7 above)
Comment (akaszynski) [reply pyvista/plotting/plotting.py]: Implemented in 32fcae6f21b8213e6c988a2d78a960e9a0300d09.

File: pyvista/plotting/mapper.py (hunk 6 above)
Comment (akaszynski) [reply pyvista/plotting/mapper.py]: Resolved in 32fcae6f21b8213e6c988a2d78a960e9a0300d09.

File: pyvista/plotting/plotting.py [hunk 10]
```diff
@@ -4585,6 +4595,61 @@ def export_obj(self, filename):
...
+    def _remove_added_scalars(self):
+        """Remove any scalars added to this plotter's datasets by this plotter.
+
+        Additional point or cell data may be added to be able to correctly plot
+        scalars. This usually occurs in one of the following cases:
+
+        * ``<scalar-name>-real`` for complex data
+        * ``<scalar-name>-normed`` for normalized multi-component arrays.
+        * ``scalars`` when adding scalars within ``add_mesh`` rather than
+          selecting an existing array.
```
Comment (adeak) [inline pyvista/plotting/plotting.py:4621]: Thanks to https://github.com/pyvista/pyvista/pull/2818 I came across another case: when supplying opacity through an array, VTK generates `__custom_rgba`, and even `__custom_rgba-real` if the scalars happened to be complex.

I don't know if we want to mention this case too, but it's probably not a case for which we can reactivate other scalars, so we don't have to change the logic because of this. I'm sure there are other weird examples too, so this need not be comprehensive (and doesn't even claim to be that).

I guess there's also the <scalar-name>-<index> case which is not mentioned here either.

**Review by adeak — APPROVED** [review]
This looks good to me now, with one remark. I've also changed a few things again, please review.

File: pyvista/plotting/plotting.py (hunk 10 above)
Comment (akaszynski) [reply pyvista/plotting/plotting.py]: > it's probably not a case for which we can reactivate other scalars

Sadly, not yet. We could add that logic in, but there isn't a clean way to do this and it's not enough of an issue for me to invest the time to add it in.

Added the note about `__custom_rgba` and the component array in c24f4abe186fc0d0faad0dedc2dbc9f8aa9c8187.

File: pyvista/plotting/plotting.py [hunk 11]
```diff
@@ -4585,6 +4595,65 @@ def export_obj(self, filename):
...
+    def _remove_added_scalars(self):
+        """Remove any scalars added to this plotter's datasets by this plotter.
+
+        Additional point or cell data may be added to be able to correctly plot
+        scalars. This usually occurs in one of the following cases:
+
+        * ``<scalar-name>-real`` for complex data
+        * ``<scalar-name>-normed`` for normalized multi-component arrays.
+        * ``scalars-<index>`` when setting ``component=<index>`` in
+          ``add_mesh`` when plotting multi-component arrays.
```
Comment (adeak) [inline pyvista/plotting/plotting.py:4621]: Just to follow the previous general cases
```suggestion
        * ``<scalars-name>-<index>`` when setting ``component=<index>`` in
          ``add_mesh`` when plotting multi-component arrays.
```

File: pyvista/plotting/plotting.py [hunk 12]
```diff
@@ -4585,6 +4595,65 @@ def export_obj(self, filename):
...
+        * ``scalars`` when adding scalars within ``add_mesh`` rather than
+          selecting an existing array.
```
Comment (adeak) [inline pyvista/plotting/plotting.py:4623]: This dummy name seems to be `"Data"`, not `"scalars"`. Or am I missing the intent here?

File: tests/test_plotter.py [hunk 13]
```diff
@@ -90,3 +90,107 @@ def test_prepare_smooth_shading_not_poly(hexbeam):
...
+
+    # however, the original active array should remain active
+    assert sphere.point_data.active_scalars_name == point_data_name
+
+
+def test_remove_scalars_rgba(sphere):
+    point_data_name = 'data'
+    sphere[point_data_name] = np.random.random(sphere.n_points)
+    pl = pyvista.Plotter()
+    pl.add_mesh(sphere, scalars=point_data_name, opacity='data')
```
Comment (adeak) [inline tests/test_plotter.py:187]: Perhaps clearer:
```suggestion
    pl.add_mesh(sphere, scalars=point_data_name, opacity=point_data_name)
```

**Review by adeak — COMMENTED** [review]
Looks almost perfect to me, just a few last nitpicks.

File: pyvista/plotting/plotting.py (hunk 12 above)
Comment (akaszynski) [reply pyvista/plotting/plotting.py:4623]:
```suggestion
        * ``Data`` when passing an array as ``scalars`` with ``add_mesh`` rather than
          selecting an existing array using a string.
```

**Review by adeak — APPROVED** [review]
LGTM, thanks!
